core/trino-main/src/main/java/io/trino/operator/output/columnar/PositionsAppenderFactory.java

```diff
@@ -0,0 +1,297 @@
+        try {
+            return isolatedAppenderClass.getConstructor(Type.class).newInstance(type);
+        }
+        catch (ReflectiveOperationException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private PositionsAppender isolateAppender(Class<? extends PositionsAppender> appenderClass)
+    {
+        Class<? extends PositionsAppender> isolatedAppenderClass = isolateAppenderClass(appenderClass);
+        try {
+            return isolatedAppenderClass.getConstructor().newInstance();
+        }
+        catch (ReflectiveOperationException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private Class<? extends PositionsAppender> isolateAppenderClass(Class<? extends PositionsAppender> appenderClass)
+    {
+        DynamicClassLoader dynamicClassLoader = new DynamicClassLoader(PositionsAppender.class.getClassLoader());
+
+        Class<? extends PositionsAppender> isolatedBatchPositionsTransferClass = IsolatedClass.isolateClass(
+                dynamicClassLoader,
+                PositionsAppender.class,
+                appenderClass);
+        return isolatedBatchPositionsTransferClass;
+    }
+
+    public static class LongPositionsAppender
+            implements PositionsAppender
+    {
+        @Override
+        public void appendTo(IntArrayList positions, Block block, BlockBuilder blockBuilder)
+        {
+            int[] positionArray = positions.elements();
+            for (int i = 0; i < positions.size(); i++) {
+                int position = positionArray[i];
+                if (block.isNull(position)) {
+                    blockBuilder.appendNull();
+                }
+                else {
+                    blockBuilder.writeLong(block.getLong(position, 0)).closeEntry();
+                }
+            }
+        }
+    }
+
+    public static class IntPositionsAppender
+            implements PositionsAppender
+    {
+        @Override
+        public void appendTo(IntArrayList positions, Block block, BlockBuilder blockBuilder)
+        {
+            int[] positionArray = positions.elements();
+            for (int i = 0; i < positions.size(); i++) {
+                int position = positionArray[i];
+                if (block.isNull(position)) {
+                    blockBuilder.appendNull();
+                }
+                else {
+                    blockBuilder.writeInt(block.getInt(position, 0)).closeEntry();
+                }
+            }
+        }
+    }
+
+    public static class BytePositionsAppender
+            implements PositionsAppender
+    {
+        @Override
+        public void appendTo(IntArrayList positions, Block block, BlockBuilder blockBuilder)
+        {
+            int[] positionArray = positions.elements();
+            for (int i = 0; i < positions.size(); i++) {
+                int position = positionArray[i];
```

Comment:
what about handling `mayHaveNull` here and in other places?

Reply:
added separate commit (could be squshed if we decide to keep it). Seems to help relatively a lot in tpcds/tpch
[cpoo_sf1000_orc_part_mayHaveNull.pdf](https://github.com/trinodb/trino/files/8173478/cpoo_sf1000_orc_part_mayHaveNull.pdf)